dual_net.py

```diff
@@ -321,58 +193,133 @@ def my_res_layer(inputs, train_mode):
-        'logits': logits,
-        'policy_output': policy_output,
-        'value_output': value_output,
-    }
 
-
-def train_ops(input_tensors, output_tensors, **hparams):
-    global_step = tf.Variable(0, name="global_step", trainable=False)
+    # train ops
+    global_step = tf.train.get_or_create_global_step()
     policy_cost = tf.reduce_mean(
         tf.nn.softmax_cross_entropy_with_logits(
-            logits=output_tensors['logits'],
-            labels=input_tensors['pi_tensor']))
-    value_cost = tf.reduce_mean(tf.square(
-        output_tensors['value_output'] - input_tensors['value_tensor']))
-    l2_cost = hparams['l2_strength'] * tf.add_n([tf.nn.l2_loss(v)
-                                                 for v in tf.trainable_variables() if not 'bias' in v.name])
+            logits=logits, labels=labels['pi_tensor']))
+    value_cost = tf.reduce_mean(
+        tf.square(value_output - labels['value_tensor']))
+    l2_cost = params['l2_strength'] * tf.add_n([tf.nn.l2_loss(v)
+        for v in tf.trainable_variables() if not 'bias' in v.name])
     combined_cost = policy_cost + value_cost + l2_cost
-    policy_output = output_tensors['policy_output']
     policy_entropy = -tf.reduce_mean(tf.reduce_sum(
         policy_output * tf.log(policy_output), axis=1))
-    boundaries = list(map(int, [1e6, 2 * 1e6]))
+    boundaries = [int(1e6), int(2e6)]
     values = [1e-2, 1e-3, 1e-4]
     learning_rate = tf.train.piecewise_constant(
         global_step, boundaries, values)
     update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
     with tf.control_dependencies(update_ops):
         train_op = tf.train.MomentumOptimizer(
-            learning_rate, hparams['momentum']).minimize(
+            learning_rate, params['momentum']).minimize(
                 combined_cost, global_step=global_step)
 
-    return {
-        'policy_cost': policy_cost,
-        'value_cost': value_cost,
-        'l2_cost': l2_cost,
-        'combined_cost': combined_cost,
-        'policy_entropy': policy_entropy,
-        'global_step': global_step,
-        'train_op': train_op,
-    }
-
 
-def logging_ops():
-    return tf.summary.merge([
-        tf.summary.histogram(weight_var.name, weight_var)
-        for weight_var in tf.trainable_variables()],
-        name="weight_summaries")
+    metric_ops = {
+        'accuracy': tf.metrics.accuracy(labels=labels['pi_tensor'],
+                                        predictions=policy_output,
+                                        name='accuracy_op'),
+        'policy_cost': tf.metrics.mean(policy_cost),
+        'value_cost': tf.metrics.mean(value_cost),
+        'l2_cost': tf.metrics.mean(l2_cost),
+        'policy_entropy': tf.metrics.mean(policy_entropy),
+        'combined_cost': tf.metrics.mean(combined_cost),
+    }
+    # Register metrics as tf.summary to make them show up during training
+    for metric_name, metric_op in metric_ops.items():
+        tf.summary.scalar(metric_name, metric_op[1])
```

Comment:
[1] is the op, i guess?  Seems like tf should use a namedtuple or something.

Reply:
Digging into it, `metric_op` is a tuple of
mean: A Tensor representing the current mean, the value of total divided by count.
update_op: An operation that increments the total and count variables appropriately and whose value matches mean_value.

It seems like the only difference is whether all N total/count are used, or N-1 total/count are used. Agree it's still a bit confusing.